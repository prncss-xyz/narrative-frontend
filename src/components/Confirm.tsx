import { ReactNode } from "react";
import { Flex, OverlayActionBox } from "./basics";
import { Clickable } from "./Clickable";
import { Overlay } from "./Overlay";

/**
 * Set an overlay dialogue with custom message to answer a yes/no question.
 * Clicking yes calls the handler, clicking outside or no close the dialogue.
 */
export function Confirm({
  overlayVisible,
  setOverlayVisible,
  handler,
  children,
}: {
  overlayVisible: boolean;
  setOverlayVisible: (state: boolean) => void;
  handler: () => void;
  children: ReactNode;
}) {
  return (
    <Overlay
      visible={overlayVisible}
      setVisible={() => setOverlayVisible(false)}
    >
      {children}
      <Flex mt={4} gap={2} justifyContent="flex-end">
        <Clickable onClick={handler}>
          <OverlayActionBox>Yes</OverlayActionBox>
        </Clickable>
        <Clickable onClick={() => setOverlayVisible(false)}>
          <OverlayActionBox>No</OverlayActionBox>
        </Clickable>
      </Flex>
    </Overlay>
  );
}
