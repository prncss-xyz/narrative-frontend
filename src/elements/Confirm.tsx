import { useTheme } from "@emotion/react";
import { ReactNode } from "react";
import { Clickable } from "./Clickable";
import { Overlay } from "./Overlay";

export function Confirm({
  overlayVisible,
  setOverlayVisible,
  handler,
  children,
}: {
  overlayVisible: boolean;
  setOverlayVisible: (state: boolean) => void;
  handler: () => void;
  children: ReactNode;
}) {
  const theme = useTheme() as any;
  return (
    <Overlay
      visible={overlayVisible}
      onClickOutside={() => setOverlayVisible(false)}
    >
      {children}
      <Flex
        mt={4}
        css={{
          justifyContent: "flex-end",
          gap: theme.space[2],
        }}
      >
        <Clickable onClick={handler}>
          <Box css={theme.styles.overlayActionButton}>Yes</Box>
        </Clickable>
        <Clickable onClick={() => setOverlayVisible(false)}>
          <Box css={theme.styles.overlayActionButton}>No</Box>
        </Clickable>
      </Flex>
    </Overlay>
  );
}
import { Box, Flex } from "./shared";
