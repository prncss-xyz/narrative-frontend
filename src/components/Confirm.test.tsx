import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Confirm } from "./Confirm";

describe("Confirm", () => {
  describe("closed", () => {
    const { container } = render(
      <div data-testid="closed-outside">
        <Confirm
          overlayVisible={false}
          setOverlayVisible={() => {
            /* do nothing */
          }}
          handler={() => {
            /* do nothing */
          }}
        >
          <div>closed-inside</div>
        </Confirm>
      </div>
    );
    it("should not find inside", () => {
      expect(() => {
        screen.getByText("closed-inside");
      }).toThrowErrorMatchingSnapshot('"error"');
    });
    it("should match snapshot", () => {
      expect(container).toMatchSnapshot();
    });
  });
  describe("opened", async () => {
    const setOverlayVisible = vi.fn();
    const handler = vi.fn();
    const { container } = render(
      <div data-testid="opened-outside">
        <Confirm
          overlayVisible={true}
          setOverlayVisible={setOverlayVisible}
          handler={handler}
        >
          <div>opened-inside</div>
        </Confirm>
      </div>
    );
    await userEvent.click(screen.getByText("Yes"));
    it("should call handler", () => {
      expect(handler).toBeCalledTimes(1);
    });
    await userEvent.click(screen.getByText("No"));
    it("should call setOverlayVisible", () => {
      expect(setOverlayVisible).toBeCalledTimes(1);
      expect(setOverlayVisible).toBeCalledWith(false);
    });
    it("should match snapshot", () => {
      expect(container).toMatchSnapshot();
    });
  });
});
