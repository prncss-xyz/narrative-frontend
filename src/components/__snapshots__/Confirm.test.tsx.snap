// Vitest Snapshot v1

exports[`Confirm > closed > should match snapshot 1`] = `<div />`;

exports[`Confirm > closed > should not find inside > "error" 1`] = `
"Unable to find an element with the text: closed-inside. This could be because the text is broken up by multiple elements. In this case, you can provide a function for your text matcher to make your matcher more flexible.

Ignored nodes: comments, script, style
[36m<body>[39m
  [36m<div>[39m
    [36m<div[39m
      [33mdata-testid[39m=[32m\\"closed-outside\\"[39m
    [36m/>[39m
  [36m</div>[39m
  [36m<div>[39m
    [36m<div[39m
      [33mdata-testid[39m=[32m\\"opened-outside\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"css-5uqfag-Box-Overlay e2i8k6n3\\"[39m
        [33mheight[39m=[32m\\"100vh\\"[39m
        [33mwidth[39m=[32m\\"100vw\\"[39m
      [36m/>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"css-1q8bug2-Box-Flex-Overlay e2i8k6n2\\"[39m
        [33mheight[39m=[32m\\"100vh\\"[39m
        [33mwidth[39m=[32m\\"100vw\\"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m\\"css-18ya61d-Box-Flex e2i8k6n2\\"[39m
          [33mheight[39m=[32m\\"100%\\"[39m
          [33mwidth[39m=[32m\\"100%\\"[39m
        [36m>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"css-15qjs6w-Box-Flex e2i8k6n2\\"[39m
            [33mcolor[39m=[32m\\"text\\"[39m
          [36m>[39m
            [36m<div[39m
              [33mclass[39m=[32m\\"css-191d8jn-Box e2i8k6n3\\"[39m
            [36m>[39m
              [36m<div>[39m
                [0mopened-inside[0m
              [36m</div>[39m
              [36m<div[39m
                [33mclass[39m=[32m\\"css-1ne0ra-Box-Flex e2i8k6n2\\"[39m
              [36m>[39m
                [36m<div[39m
                  [33mclass[39m=[32m\\"css-1x3sxzl-Box e2i8k6n3\\"[39m
                [36m>[39m
                  [36m<button[39m
                    [33mclass[39m=[32m\\"css-19fpo8m-Box e2i8k6n3\\"[39m
                    [33mcursor[39m=[32m\\"pointer\\"[39m
                    [33mheight[39m=[32m\\"100%\\"[39m
                  [36m>[39m
                    [36m<div[39m
                      [33mclass[39m=[32m\\"css-bgepeq-Box e2i8k6n3\\"[39m
                      [33mcolor[39m=[32m\\"tone4\\"[39m
                    [36m>[39m
                      [0mYes[0m
                    [36m</div>[39m
                  [36m</button>[39m
                [36m</div>[39m
                [36m<div[39m
                  [33mclass[39m=[32m\\"css-1x3sxzl-Box e2i8k6n3\\"[39m
                [36m>[39m
                  [36m<button[39m
                    [33mclass[39m=[32m\\"css-19fpo8m-Box e2i8k6n3\\"[39m
                    [33mcursor[39m=[32m\\"pointer\\"[39m
                    [33mheight[39m=[32m\\"100%\\"[39m
                  [36m>[39m
                    [36m<div[39m
                      [33mclass[39m=[32m\\"css-bgepeq-Box e2i8k6n3\\"[39m
                      [33mcolor[39m=[32m\\"tone4\\"[39m
                    [36m>[39m
                      [0mNo[0m
                    [36m</div>[39m
                  [36m</button>[39m
                [36m</div>[39m
              [36m</div>[39m
            [36m</div>[39m
          [36m</div>[39m
        [36m</div>[39m
      [36m</div>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</body>[39m"
`;

exports[`Confirm > opened > should match snapshot 1`] = `<div />`;
